Dynatrace/dynatrace-operator
# Let troubleshooting command continue execution until unrecoverable error is found

# Description
Currently, troubleshoot command executes series of `tests` on operator deployment. If one of them fails, `troubleshoot` exists early reporting first problem found.

Changes in scope of this PR:
- renames `test` to `check` to avoid confusion with testing code
- `checks` execution is moved to extracted function `runChecks` [1]
- `Check` struct is used to specify check including prerequisites required to run it
- `error`s wrapping is fixed (no double printing or pseudo-json structure)
- `runChecks` returns `AggregatedError`. It's implementation is troubleshoot-specific, shouldn't be used elsewhere in this form. Meanwhile, waiting for https://github.com/golang/go/issues/53435 🥺

[1]
`Check` struct contains prerequisites in form of a slice of strings containing pointers to other checks (directed acyclic graph).
If any of prerequisites failed or was skipped, check will be skipped.

## How can this be tested?
Run `k exec deploy/dynatrace-operator -n dynatrace -- dynatrace-operator troubleshoot` on your environment.
Try deleting e.g. some secrets to reproduce scenario with multiple problems existence.

Example output:
```
kubectl exec deploy/dynatrace-operator -n dynatrace -- dynatrace-operator troubleshoot --dynakube dk
{"level":"info","ts":"2022-11-15T14:20:56.184Z","logger":"version","msg":"dynatrace-operator","version":"snapshot-feature-continue-troubleshooting-on-non-crit-error","gitCommit":"a86866db8264690eb4694e9525675331f09bf67d","buildDate":"2022-11-15T14:15:27+00:00","goVersion":"go1.19.3","platform":"linux/amd64"}
[namespace ]    --- checking if namespace 'dynatrace' exists ...
[namespace ]     ✓  using namespace 'dynatrace'
[dynakube  ]    --- checking if 'dynatrace:dk' Dynakube is configured correctly
[dynakube  ]        CRD for Dynakube exists
[dynakube  ]        using 'dynatrace:dk' Dynakube
[dynakube  ]        checking if api url is valid
[dynakube  ]        api url is valid
[dynakube  ]     X  'dynatrace:dk' secret is missing: secrets "dk" not found
[dynakube  ]     ⚠  Skipped 'dynatraceApiSecretHasApiToken' check because prerequisites aren't met: [apiSecret]
[dynakube  ]     X  'dynatrace:dk-pull-secret' pull secret is missing: secrets "dk-pull-secret" not found
[dynakube  ]     ⚠  Skipped 'pullSecretHasRequiredTokens' check because prerequisites aren't met: [pullSecretExists]
[dynakube  ]        no proxy is configured
[dynakube  ]     X  'dynatrace:dk' Dynakube isn't valid. Target namespace and dynakube can be changed by providing '--namespace <namespace>' or '--dynakube <dynakube>' parameters.: 'dynatrace:dk' secret is missing: secrets "dk" not found
'dynatrace:dk-pull-secret' pull secret is missing: secrets "dk-pull-secret" not found
[dynakube  ]     ⚠  Skipped 'DynatraceClusterConnection' check because prerequisites aren't met: [dynakube]
[dynakube  ]     ⚠  Skipped 'imagePullable' check because prerequisites aren't met: [dynakube]
```


## Checklist
